### PR TITLE
Improve the error message for unsupported Buffer in Buffer situation.

### PR DESCRIPTION
**Pull request by stuartarchibald**
As title.

xref: #6473

### CODE REVIEW

File: numba/core/types/common.py [hunk 1]
```diff
@@ -49,7 +49,10 @@ def __init__(self, dtype, ndim, layout, readonly=False, name=None):
         from .misc import unliteral
 
         if isinstance(dtype, Buffer):
-            raise TypeError("Buffer dtype cannot be buffer, have dtype: {}".format(dtype))
+            msg = ("The dtype of a Buffer type cannot itself be a Buffer type, "
+                   "this is unsupported behaviour."
+                   "\nThe dtype requested for the offending Buffer was: {}.")
```
Comment (seibert) [inline numba/core/types/common.py:54]: maybe switch "offending" to "unsupported"?  Seems less harsh.  :)

Comment (stuartarchibald) [inline numba/core/types/common.py:54]:
```suggestion
                   "\nThe dtype requested for the unsupported Buffer was: {}.")
```

File: numba/tests/test_record_dtype.py [hunk 2]
```diff
@@ -1303,5 +1303,24 @@ def foo(rec, flag=None):
...
+        @njit()
+        def foo(x):
+            x["y"][0] = 1
+
+        dt = np.dtype([("y", (np.uint64, 5)),])
+        x = np.ones(1, dtype=dt)
+        with self.assertRaises(TypingError) as e:
+            foo(x)
+        ex1 = "The dtype of a Buffer type cannot itself be a Buffer type"
+        ex2 = "offending Buffer was: nestedarray(uint64, (5,))"
```
Comment (stuartarchibald) [inline numba/tests/test_record_dtype.py:1319]:
```suggestion
        ex2 = "unsupported Buffer was: nestedarray(uint64, (5,))"
```

File: numba/core/types/common.py (hunk 1 above)
Comment (stuartarchibald) [reply numba/core/types/common.py]: thanks, done in 
d68d688

**Review by seibert — APPROVED** [review]
looks good!

**Pull request merged**